data-prepper-plugins/cloudwatch-logs/src/main/java/org/opensearch/dataprepper/plugins/sink/client/CloudWatchLogsService.java

```diff
@@ -0,0 +1,152 @@
+
+    public CloudWatchLogsService(final Buffer buffer,
+                                 final CloudWatchLogsClient cloudWatchLogsClient,
+                                 final CloudWatchLogsMetrics cloudWatchLogsMetrics,
+                                 final CloudWatchLogsLimits cloudWatchLogsLimits,
+                                 final BlockingQueue<ThreadTaskEvents> blockingQueue,
+                                 final String logGroup, final String logStream,
+                                 final int retryCount, final long backOffTimeBase) {
+
+        this.buffer = buffer;
+        this.cloudWatchLogsClient = cloudWatchLogsClient;
+        this.cloudWatchLogsMetrics = cloudWatchLogsMetrics;
+        this.cloudWatchLogsLimits = cloudWatchLogsLimits;
+        this.taskQueue = blockingQueue;
+        this.logGroup = logGroup;
+        this.logStream = logStream;
+        this.retryCount = retryCount;
+        this.backOffTimeBase = backOffTimeBase;
+
+        this.bufferedEventHandles = new ArrayList<>();
+
+        bufferLock = new ReentrantLock();
+        sinkStopWatch = new SinkStopWatch();
+        asyncExecutor = newCachedThreadPool();
+    }
+
+    /**
+     * Function handles the packaging of events into log events before sending a bulk request to CloudWatchLogs.
+     * Implements simple conditional buffer. (Sends once batch size, request size in bytes, or time limit is reached)
+     * @param logs - Collection of Record events which hold log data.
+     */
+    public void processLogEvents(final Collection<Record<Event>> logs) {
+        sinkStopWatch.startIfNotRunning();
+        for (Record<Event> log: logs) {
+            String logString = log.getData().toJsonString();
+            int logLength = logString.length();
+
+            if (cloudWatchLogsLimits.isGreaterThanMaxEventSize(logLength)) {
+                LOG.warn("Event blocked due to Max Size restriction! {Event Size: {} bytes}", (logLength + CloudWatchLogsLimits.APPROXIMATE_LOG_EVENT_OVERHEAD_SIZE));
+                continue;
+            }
+
+            long time = sinkStopWatch.getStopWatchTimeSeconds();
+
+            bufferLock.lock();
+
+            int bufferSize = buffer.getBufferSize();
+            int bufferEventCount = buffer.getEventCount();
+            int bufferEventCountWithEvent = bufferEventCount + 1;
+            int bufferSizeWithAddedEvent = bufferSize + logLength;
+
+            if ((cloudWatchLogsLimits.isGreaterThanLimitReached(time, bufferSizeWithAddedEvent, bufferEventCountWithEvent) && (bufferEventCount > 0))) {
+                stageLogEvents();
+                addToBuffer(log, logString);
+            } else if (cloudWatchLogsLimits.isEqualToLimitReached(bufferSizeWithAddedEvent, bufferEventCountWithEvent)) {
+                addToBuffer(log, logString);
+                stageLogEvents();
+            } else {
+                addToBuffer(log, logString);
+            }
+
+            bufferLock.unlock();
```

Comment:
please put behind `finally` block
https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/locks/ReentrantLock.html

Reply:
Thank you! I missed this while reading the documentation, I have made the change in the latest revision.